Datum Emotes MK II

VampyrBytes opened the pull request:
Revamp (excuse the pun) of the emote system. Anyone who's been in the current system will know how horrible it is with procs up to 800+ lines long, and code copy pastad both within procs and between the various emote.dm's.

#### Features
* Far more oop
* Allows different types of mobs to access the same emote rather than having to copy code
* Emotes can be overridden for different types of mob, only having to change the bits that are different
* Automatically searches for the appropriate version of an emote for a mob
* Help emote automatically shows available emote commands rather than relying on the emote creator adding it to the help emote
 * Emotes now accessed through verbs as well as *say
 * Emote parameters are now taken through input boxes instead of breaking up the say string
 * If an emote fails, it tells you why it failed
 * Personalised messages. eg if Fox McCloud hugs Kiera Williams, Most people will still see the normal message, but Fox will see 'You hug Kiera Williams' and I will see 'Fox McCloud hugs you'. Also changes text macros appropriately - eg \his shows as your in your version of the message
 * Support for blind and deaf messages. By default with audible messages a blind person will know what happened but not who did it, and a deaf person will get the mime version of the emote
 * All default behaviour easily overridable, so if it's not appropriate for a particular emote it's easy to change

#### Details
Only one instance of each emote datum exists, held in a global list. Every mob has an emoteHandler datum that allows access to the appropriate emotes. Each verb creates a dummy object in the user's contents to give it a valid loc to work from

#### Gameplay
Only a few minor tweaks to actual emotes made in this PR - I'm planning another after this to deal with inconsistencies and add features to emotes
 * NPC monkeys were trying to perform emotes that they didn't have access to, they now have access to all the emotes they need
 * I combined the old flip with the new one, so now if you have a target and have a grab on them you flip over them, but if you have a target and don't have a grab on them you flip in their general direction
 * Whilst most emotes that are affected by being muzzled led you to making some kind of noise, a few didn't - now all do
I think that's all the gameplay changes, but this PR has been going for some time, and it's possible I've forgotten some... if there are any more they are minor and more to do with consistancy than intending to change the emote

I'm not sure whether this does solve the problem in #4268 - I can't test it locally as you need a guardian to hear the emote and an owner to perform the emote, so it'll need testing after once this is merged

:cl: VampyrBytes
rscadd: Emotes can now be accessed through verbs as well as say *
rscdel: Emotes accessed through say * no longer take parameters after a -
rscadd: Emotes now take parameters through input boxes
rscadd: Emotes now have support for blind and deaf characters
rscadd: Emotes now personalise messages
tweak: Flip can now be targeted with or without a grab
tweak: All emotes that are affected by being muzzled now make the user make some kind of noise
bugfix: monkeys now have access to all the emotes that the npc tries to perform
/ :cl:

Glorken commented:
Sounds great except for the removal from say. With hotkeys on, pressing 'T' and an asterisk is much more convenient than going through a tab and finding the verb or typing the verb into the bar at the bottom.

VampyrBytes commented:
@Glorken 
Thanks for the input. I tend not to use hotkeys, so this was just the type of thing I needed to know :smile: 
TBH my main issue with it going through say was with stripping the parameter out of the text, which was messy and limited emotes to having one parameter. With the introduction of verbs I need to change how parameters are taken anyway (hopefully I should have an update for this later today) so leaving in the ability to access them through say * isn't an issue.

VampyrBytes commented:
@TheDZD 
You missed my work in progress and do not merge labels :wink:

VampyrBytes commented:
Ok, parameters are now taken in through input boxes (with a cancel option). There's built in support for a target (eg a mob) and one other parameter, but overriding could let you take more.
So far there's support for Mobs or Atoms as the target, and a number as the other parameter, as they're ones I've come across, but adding more support should be easy and I'll do it as I convert emotes that need others.
For testing I added the signal emote (raises x fingers) and gave it the ability to take a mob target (at someone) :wink: I can remove it having the target if people prefer :smile: 

Anyone have issues with parameters being taken in through input boxes?

optimumtact commented:
/me lurks

VampyrBytes commented:
It's getting there now, just a load of testing to do and the old code to strip out

Fox-McCloud commented:
Status?

VampyrBytes commented:
@Fox-McCloud 
Still going - slowly because of rl stuff. Got a fair amount of testing done, and some bugs fixed locally, Working on a couple of bugs with the ghost custom emote atm, which are hopefully the last ones.

Krausus commented:
It might be a little early to critique this, seeing as it's still WIP, but...

```
/datum/emote/hug/available(var/mob/user)
	if(ishuman(user))
		return 1
```

There are 39 emotes where this is the entirety of their `available` check. Excessive copy-paste. You could get rid of 38 of those by just putting it on, say, a `humanonly` subtype, and making them subtypes of that. In fact, I'm surprised by just how little subtyping there is, considering that's one of the biggest benefits of using datums.

On a related note, there really needs to be a better way to specify emote availability. Dozens of `iswhatever` checks are *awful.* Maybe, like, a set of flags? Hmm.

```
/datum/emote/hug/New()
	..()
	commands += "hug"
	commands += "hugs"
```

Almost every `New()` is like this. I'm pretty sure the overwhelming majority of emotes could just have something like `commands = list("hug", "hugs")` in their vars with no loss of functionality. I know this method is meant to *add* commands instead of *replacing* them, but does that actually serve any useful function in this system's current state?

> Anyone have issues with parameters being taken in through input boxes?

Yes. Having an input box pop up every time you do an emote that could potentially be targeted at someone sounds incredibly inconvenient, and feels like a step backwards. The reason targeting emotes in the current system sucks is because target names are expected to be *exact.* Having targeting logic handled in a single place is a golden opportunity to implement fuzzy name matching.

VampyrBytes commented:
> It might be a little early to critique this, seeing as it's still WIP, but...

> 

This is pretty close to finished now, so critique is welcome

> There are 39 emotes where this is the entirety of their available check. Excessive copy-paste. You could get rid of 38 of those by just putting it on, say, a humanonly subtype, and making them subtypes of that. In fact, I'm surprised by just how little subtyping there is, considering that's one of the biggest benefits of using datums.

> 
I am very reluctant to start subtyping by who can use the emote as that is what led to the current mess. I do agree there could probably be better subtyping, but I'd be far more likely to go with subtyping by type of emote (eg audible/visual) than with who can access it, which is something that can change over time.

> On a related note, there really needs to be a better way to specify emote availability. Dozens of iswhatever checks are awful. Maybe, like, a set of flags? Hmm.

> 
Taking the emotes out of the mob tree means that at some point the iswhatever check has to be done. Yes, I could implement flags which would remove the copypasta, but you're going to end up with a proc that looks like
`/datum/emote/available(var/mob/user)
if(flags & human)
     if(ishuman(user))
           return 1
if(flags & larva)
     if(islarva(user))
           return 1`
and so on for every possibility... which means it's got to run through every if in that proc everytime. Bear in mind that every mob(player, npc, living, dead) has an emote handler and every emote handler does a recursive search through the emotes to find the available ones when it's created, so the last thing you want at round start is for each available check to have a dozen or more if statements to go through.

> /datum/emote/hug/New()
    ..()
    commands += "hug"
    commands += "hugs"

Almost every New() is like this. I'm pretty sure the overwhelming majority of emotes could just have something like commands = list("hug", "hugs") in their vars with no loss of functionality. I know this method is meant to add commands instead of replacing them, but does that actually serve any useful function in this system's current state?

> 
Thank you, that is soooo much nicer. I'll definitely be doing this

> Yes. Having an input box pop up every time you do an emote that could potentially be targeted at someone sounds incredibly inconvenient, and feels like a step backwards. The reason targeting emotes in the current system sucks is because target names are expected to be exact. Having targeting logic handled in a single place is a golden opportunity to implement fuzzy name matching.

> 
That is one reason the current targeting system sucks. Another is that you have to know whether an emote takes an input and what kind it takes (see #4596 ). Yet another is that it limits emotes to only taking one input. Oh, and you've got to know the name of the target, you can't just be passing someone and emote at them without figuring out their name, then typing it, by which time they're long gone. I'm also unclear on how the input box would be 'incredibly inconvenient'?

Krausus commented:
> I am very reluctant to start subtyping by who can use the emote as that is what led to the current mess.

What led to the current mess was a mixture of lazily bolting on new emotes and lazily copy-pasting to support different emoters. I wouldn't subtype by mob type either, but that's basically what many emotes are already doing anyway.

>Yes, I could implement flags which would remove the copypasta, but you're going to end up with a proc that looks like
`/datum/emote/available(var/mob/user)
if(flags & human)
if(ishuman(user))

You certainly *could* cut down on the checks like that, but you're still doing `iswhatever` checks. I was thinking more in terms of flags, stored on mobs and on emotes, representing what features a mob can emote with. For example, you might have a "has arms" flag, and the hug emote would require that flag; humans, adult aliens, space bears, etc. would have that flag set, and could hug, whereas something like a space carp wouldn't be able to hug, because it doesn't have arms and is a merciless killing machine. Something like that would make it trivial to add a new mob and quickly define what traits give it access to what emotes.

> the last thing you want at round start is for each available check to have a dozen or more if statements to go through.

You are **vastly** underestimating how quickly BYOND can chew through a few hundred if statements.

> you have to know whether an emote takes an input and what kind it takes (see #4596)

Sure, but there's other ways for players to get that information - the list of emotes, descriptions on the verbs, trial and error, etc. By the way, you're pointing to an example where the *real* issue was the lack of feedback on failure, not the player not knowing how the emote works.

> Yet another is that it limits emotes to only taking one input

Why would emotes be limited to a single input? Because names can have spaces in them? Have the name come after the other inputs, or just ignore spaces and match by first name.

> you've got to know the name of the target, you can't just be passing someone and emote at them without figuring out their name, then typing it, by which time they're long gone.

So, instead of hovering over them and typing the first few letters of their name, you have to select this person, whose name you don't know, from the list of valid targets for whatever emote you're using on strangers who happen to be passing by you? Sure, in this particular, contrived example, the input box method would probably be faster, assuming there are no other valid targets nearby, because it would take way too long to find out the name of this passing stranger, so you wouldn't be able to pick them from a populated list.

> I'm also unclear on how the input box would be 'incredibly inconvenient'?

A bit of an overdramatization, of course. Personally, I don't use the push-to-talk input box, because having a box pop up over/nearby my character is annoying. If half the emotes I used popped up a box asking me whether I want to target that emote at someone nearby, it would be similarly annoying. Especially if it's something you have on a hotkey, like the ever-popular *flip.

Instead of picking one or the other, though, it could be left to player preference, either as an actual game preference or simply a special way of typing in emotes (for example, "hug ?" to get an input box).

VampyrBytes commented:
> What led to the current mess was a mixture of lazily bolting on new emotes and lazily copy-pasting to support different emoters. I wouldn't subtype by mob type either, but that's basically what many emotes are already doing anyway

Agreed. But in this PR I have purposely avoided changing the functionality of the actual emotes themselves other than the odd tweak. This PR is about the system. I'm already planning another PR once this is implemented that will look at the actual emotes themselves, both in terms of the inconsistencies with who can access which emotes and adding extra functionality (like taking more than one input where appropriate) to the emotes. That's the point where I'll subtype them better - doing it now when I know there's going to be numerous changes feels kinda pointless.

> You certainly could cut down on the checks like that, but you're still doing iswhatever checks. I was thinking more in terms of flags, stored on mobs and on emotes, representing what features a mob can emote with. For example, you might have a "has arms" flag, and the hug emote would require that flag; humans, adult aliens, space bears, etc.

Ok, that makes more sense. I haven't used flags much, so was thinking in the wrong terms. However, if you think about what else these flags could be used for you'll realise what you're talking about is well outside the scope of this PR. You're really talking about replacing the `iswhatever` system, which is something I would definitely support, and I'd happily be the one to convert the emotes to a system like that, but it's not something I'm about to start in this PR. As a sidenote, imo the best way of doing something like that would actually be to expand the scope of the species datums and the give them more flags. Species was a great idea, but only giving them to humans was (imo) terrible implementation... at least all `mob/living` should have them, if not all mobs.

> You are vastly underestimating how quickly BYOND can chew through a few hundred if statements.

Probably, but whilst I'm all for minimising/removing copypasta the extra processing required in this case just seems wasteful

> Sure, but there's other ways for players to get that information - the list of emotes, descriptions on the verbs, trial and error, etc. By the way, you're pointing to an example where the real issue was the lack of feedback on failure, not the player not knowing how the emote works.

It was a combination of both... the user didn't know *johnny took a mob parameter and the emote didn't tell the user why it had failed (something I've addressed in this system).

> Why would emotes be limited to a single input? Because names can have spaces in them? Have the name come after the other inputs, or just ignore spaces and match by first name.

Matching by first name is not possible... first names are not unique. It may sound unlikely for you to be emoting with more than one person around you with the same name, until you think about the fact that some species have their names picked for them from a list. Worse still, every adult alien's first name is alien as they are named 'alien randomNumber'. 

Having the name come last still limits the inputs you can take. It means you can only have one input that gets an object by its name as any object name can be more than one word.

Imo, breaking up a string into different things, even just the emote and a single parameter, is always going to get messy. It's the main reason I moved away from that with this system. 

> So, instead of hovering over them and typing the first few letters of their name, you have to select this person, whose name you don't know, from the list of valid targets for whatever emote you're using on strangers who happen to be passing by you?

Um, you do realise that with an input box if you press the first letter of the name it will jump to the first entry with that as the first letter, and if you keep pressing it it cycles through the entries with that first letter? Then you just hit enter to make the selection. Correct me if I'm wrong, but that sounds like the functionality you're after done in a better way than would be possible with the method you're suggesting as, afaik, DM has no way of monitoring what you're typing as you type it.

> Sure, in this particular, contrived example

I understand why it could sound contrived, but I'm not a fast typer, and I swear, it has happened to me more than once

> Instead of picking one or the other, though, it could be left to player preference, either as an actual game preference or simply a special way of typing in emotes (for example, "hug ?" to get an input box).

I guess that's a possibility, although I would much rather the system used just one way as it's cleaner. However, I'm hoping to have this ready for review soon(trademark) and once that happens I'll see what the general consensus is

Krausus commented:
> I'm already planning another PR once this is implemented that will look at the actual emotes themselves, both in terms of the inconsistencies with who can access which emotes and adding extra functionality

Fair enough.

> However, if you think about what else these flags could be used for you'll realise what you're talking about is well outside the scope of this PR.

Yeah, it was more like idle daydreaming than a serious suggestion.

> Probably, but whilst I'm all for minimising/removing copypasta the extra processing required in this case just seems wasteful

Reducing copy-paste is very often done at the expense of processing time, with the major exceptions being very performance-sensitive code (for example, the lighting system). To be clear, though, I'm not saying you *should* reduce it in this particular way. 39 identical procs, though...

> Matching by first name is not possible... first names are not unique. It may sound unlikely for you to be emoting with more than one person around you with the same name, until you think about the fact that some species have their names picked for them from a list.

The majority of the time, they're unique enough...

> Worse still, every adult alien's first name is alien as they are named 'alien randomNumber'.

... except when it comes to those generic numbered mobs. It would be neat if using a number matched it to a numbered mob, though.

> Um, you do realise that with an input box if you press the first letter of the name it will jump to the first entry with that as the first letter, and if you keep pressing it it cycles through the entries with that first letter? Then you just hit enter to make the selection.

Yes, I know how input boxes work. Your example was based on wanting to emote at someone whose name you didn't have time to find out.

> I understand why it could sound contrived, but I'm not a fast typer, and I swear, it has happened to me more than once

I concede that input boxes would probably be pretty handy for slower typers. As a faster typer, I'd still prefer some way to opt out that isn't "mash escape/enter X times", but I'll stop fussing about it.

VampyrBytes commented:
> 39 identical procs, though...

I know. I guess I could get around it with something like a humanAccess var and then at the datum/emote level have 
`if(humanAccess && ishuman(user))
return 1`

but it just feels snowflakey to have a var just to give humans access. Again, I'll see what the general opinion is once it's up for review

VampyrBytes commented:
@Fox-McCloud @TheDZD 

This is (finally!) ready for review

Krausus commented:
Players in the lobby get a "Me" verb under the "Emote" tab that simply runtimes.

Using an emote with a target seems to make you face eastward afterwards, regardless of where the target is.

The target input box's "Cancel" button is used to specify no target. It would make more sense, and be more convenient, for "Cancel" to cancel the emote entirely, with no target being be an option such as "(None)" in the list, selected by default. Currently, using cancel for "no target" means pressing or tabbing over to the cancel button if you don't want a target, since the escape key doesn't cancel.

Some emotes have a period at the end of their messages, most do not.

VampyrBytes commented:
Thanks for the input Krausus :smile: 

> Players in the lobby get a "Me" verb under the "Emote" tab that simply runtimes.

It being there is just part of it being a mob/verb, and it made more sense to keep it as that as every mob has it. However, it certainly shouldn't be runtiming, I'll get a fix up for that straight away.

> Using an emote with a target seems to make you face eastward afterwards, regardless of where the target is.

I've just tested this and I can't replicate it... selecting a target doesn't change my direction. Can I get the exact steps you went through?

> The target input box's "Cancel" button is used to specify no target. It would make more sense, and be more convenient, for "Cancel" to cancel the emote entirely, with no target being be an option such as "(None)" in the list, selected by default. Currently, using cancel for "no target" means pressing or tabbing over to the cancel button if you don't want a target, since the escape key doesn't cancel.

Hmmm, from a code point of view that isn't going to be particularly pretty, and there's the issue of none not being a valid choice for some emotes so it shouldn't be in their boxes. However, I do agree that from a user pov, the behaviour your suggesting makes more sense. I'll have a look and see what I can come up with.

> Some emotes have a period at the end of their messages, most do not.

Yeah, that's my bad. I started off putting them on, then got to the first (blink iirc) that the override involved adding something to the end of the parent and realised that it was best if they didn't have them. I meant to go back and remove the others but never got around to it. I'll get rid of the others

Krausus commented:
> It being there is just part of it being a mob/verb, and it made more sense to keep it as that as every mob has it. However, it certainly shouldn't be runtiming, I'll get a fix up for that straight away.

Players in the lobby cannot speak or emote, only chat in OOC. It would be preferable for an unusable verb to not be visible.

> I've just tested this and I can't replicate it

Now that I think about it, I'm pretty sure the problem is that I was double-clicking to select from the input box, and the input box closes and shifts focus back to the main window so quickly that it registers the still-pressed mouse button as a click in the world. I remember having this issue before. BYOND's input boxes really need some improvement.

It *would* be nice if targeted emotes actually made you face your target, though. They probably already would, if the current method for handling targets wasn't such an atrocity.

> from a code point of view [handling Cancel] isn't going to be particularly pretty

It would definitely require reorganizing the code a bit, so input occurs at a point where failing out is convenient.

> there's the issue of none not being a valid choice for some emotes so it shouldn't be in their boxes

Simple. For some emotes, instead of the `canTarget = 1` they have now, have something like `mustTarget = 1`.

> I'll get rid of the others

Or, instead of outputting sentences without periods at the end, you could detect missing punctuation and add one just before you output.

VampyrBytes commented:
> Players in the lobby cannot speak or emote, only chat in OOC. It would be preferable for an unusable verb to not be visible.

Not true. Me is available from the lobby at the moment (it's under the IC tab). When you are in the lobby you are actually a mob/new_player, so you have access to any mob/verb, and I didn't create `/mob/verb/me_verb`, I just changed it to go through the emote system and to be under the emote tab. My fix actually means clicking on it does nothing, which is less than it currently does (It brings up the input box at least atm... I didn't actually try entering anything into it as I was on the live server)

> It would be nice if targeted emotes actually made you face your target, though. They probably already would, if the current method for handling targets wasn't such an atrocity.

Yeah, it would be. However it's not something I know how to do straight off, and I've got a busy few weeks ahead irl, so want to get this PR wrapped up. I'll look into it in the next emote PR.

> It would definitely require reorganizing the code a bit, so input occurs at a point where failing out is convenient.

Actually no, that's already in the system. If any input returns INVALID then it drops out. It's more that putting None in the options means manually putting the list together instead of just using `null|mob in view()` or `null|mob|obj|turf in oview`. That combined with turning null into INVALID and "None" into null fill out the procs a bit more than I'd like... the smaller the proc, the less that can go wrong, or has to be copypastad, when someone needs to override the proc. I've actually already done the coding for it, I just need to do some testing to make sure I haven't broken anything, so you'll see what I mean when I push it.

> Simple. For some emotes, instead of the canTarget = 1 they have now, have something like mustTarget = 1

Actually they have both, with mustTarget just not adding the none to the list. But yes, I even called it that lol

> Or, instead of outputting sentences without periods at the end, you could detect missing punctuation and add one just before you output.

If you look back, I'd actually already removed them before you posted. At the time I did think about adding it in in addExtras() (which deals with things like putting the spans around). The reason I didn't was because of the horrible if statement I'd have to use to check for existing punctuation, it would be something like
```
var/fullStop
var/last = getLastChar(message)
if(!(last == "." || last == "!" || last == "\"\" || last == any other punctuation marks that wouldn't be appropriate to put a . after)
     fullStop = "."
```
which is just horrible

Krausus commented:
> Me is available from the lobby at the moment (it's under the IC tab)

Oh, yeah, I forgot about that.

> The reason I didn't was because of the horrible if statement I'd have to use to check for existing punctuation, it would be something like

... like this?
```
var/regex/endingPunc = new("\[\\.!?\"]$")
if(!endingPunc.find(message))
  message += "."
```

VampyrBytes commented:
> ... like this?
```
var/regex/endingPunc = new("\[\\.!?\"]$")
if(!endingPunc.find(message))
  message += "."
```

Welp. Yeah, something like that would be much nicer :wink: Thanks. I've got a busy few days, so I'm not sure when, but I'll get this in when I get the chance.

KasparoVy commented:
Gotta love regex

VampyrBytes commented:
> Gotta love regex

In theory yes, in practice...

So far, having fixed compile times and runtimes caused by not escaping properly I've got to
```
var/regex/endPunc = new("\\.|!|\\?|\"|\\)|]|'|>")
if(!endPunc.Find(message, length(message)))
    message += "."
```

Which puts punctuation on those without just fine, unfortunately it's also removing the punctuation from any that do already have it :disappointed:
Can anyone spot where I've gone wrong?

Krausus commented:
The regular expression I gave as an example matches any one of the characters in the brackets, followed immediately by the end of the line. [Here's a visualization of it](https://regexper.com/#%5B%5C.!%3F%22%5D%24). To match additional characters, they simply need to be added in the brackets, escaping them if necessary.

Yours uses "or" pipes to accomplish the same thing brackets do, but less space-efficiently; you would normally only use those if you were matching groups. Then, instead of checking for the end of line in the regex, you only run the regex itself on the very last character. That *should* work just fine, though it's not ideal, since it's less readable and doesn't allow more complicated matching (for example, only matching parentheses with a preceding punctuation character).

> it's also removing the punctuation from any that do already have it

The code you've shown doesn't modify `message` in any way other than adding a period. It should not be possible for it to remove anything.

VampyrBytes commented:
Thanks Krausus, been a long time (read over a decade) since I've used regular expressions.

> The code you've shown doesn't modify message in any way other than adding a period. It should not be possible for it to remove anything.

Yeah, after some more testing I've realised that I put test punctuation at the end of text in some, but didn't put it in selftext, so after it went through the addextras proc it is replacing text with selftext and therefore losing the punctuation... ofc that's not an issue if the . has been put on because the . isn't in text, so it was working fine for the ones I hadn't changed to test lol

Want to see what I can do about that, and check over the other situations where a different message is used. Should be able to get it fixed up Monday, DNM in the meantime

Fox-McCloud commented on code/datums/Emote_system/emote.dm:
```diff
@@ -0,0 +1,573 @@
+				outputMessage(M, msg)
+				continue
+
+		if(M.stat == UNCONSCIOUS || (M.sleeping > 0 && M.stat != 2))
```
stat magic numbers

VampyrBytes replied:
copypasta that I didn't fix up... at least it only uses magic numbers for half of it...

not sure why the > 0 either, will fix that too

Fox-McCloud commented on code/datums/Emote_system/emotes.dm:
```diff
@@ -0,0 +1,1726 @@
+
+/datum/emote/fart/standardMessage(var/mob/user)
+	var/message
+	if(SUPER_FART in user.mutations)
```
Superfart has been 100% stripped from the emote.

VampyrBytes replied:
Ah right, I thought it had been overlooked when super farts were brought in. Will strip it out

Fox-McCloud commented on code/datums/Emote_system/emotes.dm:
```diff
@@ -0,0 +1,1726 @@
+				continue
+			if (M == user)
+				continue
+			M.reagents.add_reagent("space_drugs",rand(1,10))
```
should be jenkem; reference current emote code.

TheDZD commented on code/datums/Emote_system/emote.dm:
```diff
@@ -0,0 +1,573 @@
+		return
+	if(startText)
+		message = "[startText] [message]"
+	message = "<span class = '[emoteSpanClass]'>[message]</span>"
```
Current standard is no space between `span class`, `=` and the class. There are a lot of these.

TheDZD commented on code/datums/Emote_system/emote.dm:
```diff
@@ -0,0 +1,573 @@
+	doAction(user, params)
+
+	for (var/obj/item/weapon/implant/I in user)
+		if (I.implanted)
```
No space after if.

VampyrBytes replied:
You can tell where I copypastad lol. Will get them sorted

TheDZD commented on code/datums/Emote_system/emote_handler.dm:
```diff
@@ -0,0 +1,90 @@
+/datum/emoteHandler/proc/showCommands()
+	var/emoteList = "Available emotes are "
+	var/commandAdded = 0
+	for (var/c in commands)
```
Likewise, no space after for.

Fox-McCloud commented on code/datums/Emote_system/emotes.dm:
```diff
@@ -0,0 +1,1726 @@
+	sound = 'sound/effects/snap.ogg'
+	canTarget = 1
+	targetMob = 1
+	targetText = ""
```
keep in mind, despite being audibly, this is meant to have 0 cooldown.

TheDZD commented on code/datums/Emote_system/emote.dm:
```diff
@@ -0,0 +1,574 @@
+
+/datum/emote/proc/processMessage(var/mob/user, var/list/params, var/message = "")
+	var/visualOrAudible = audible + 1
+	testing("message = [message]")
```
Is this leftover debug code?

VampyrBytes replied:
yeah, It'll be gone in a few minutes, just doing last tests on the adding of the . if there's no ending punctuation, will be pushing final version very soon

VampyrBytes commented:
@TheDZD @Fox-McCloud 

Unless you guys can see anything else that needs doing, this is ready to go (and please merge before tiger's if ( pr, as I've no idea how to deal with merge conflicts where someone's changed a file I've deleted!)

KasparoVy commented:
Congrats dude, looks good
